planner/memo/group.go

```diff
@@ -87,15 +88,40 @@ type Group struct {
 	EngineType EngineType
 }
 
+// buildLogicalProperty builds a new LogicalProperty for the new GroupExpr.
+// it also builds the unique key info and max one row info.
+func buildLogicalProperty(e *GroupExpr, oldSchema *expression.Schema) *property.LogicalProperty {
+	prop := &property.LogicalProperty{}
+	if oldSchema == nil {
+		return prop
+	}
+	newSchema := expression.NewSchema(oldSchema.Columns...)
+	childSchema := make([]*expression.Schema, len(e.Children))
+	childMaxOneRow := make([]bool, len(e.Children))
+	for i := range e.Children {
+		childSchema[i] = e.Children[i].Prop.Schema
+		childMaxOneRow[i] = e.Children[i].Prop.MaxOneRow
+	}
+	if len(childSchema) == 1 {
+		// For UnaryPlan(such as Selection, Limit ...), we can set the child's unique key as its unique key.
+		// If the GroupExpr is a schemaProducer, schema.Keys will be reset below in `BuildKeyInfo()`.
+		newSchema.Keys = childSchema[0].Keys
+	}
+	e.ExprNode.BuildKeyInfo(newSchema, childSchema)
+	prop.Schema = newSchema
+	prop.MaxOneRow = e.ExprNode.MaxOneRow() || plannercore.HasMaxOneRow(e.ExprNode, childMaxOneRow)
+	return prop
+}
+
 // NewGroupWithSchema creates a new Group with given schema.
+// Must make sure the initial GroupExpr's children have been set.
 func NewGroupWithSchema(e *GroupExpr, s *expression.Schema) *Group {
-	prop := &property.LogicalProperty{Schema: s}
 	g := &Group{
 		Equivalents:  list.New(),
 		Fingerprints: make(map[string]*list.Element),
 		FirstExpr:    make(map[Operand]*list.Element),
 		ImplMap:      make(map[string]Implementation),
-		Prop:         prop,
+		Prop:         buildLogicalProperty(e, s),
```

Comment:
It would be better if we can make the unique key info be lazily built?

Reply:
Sounds feasible. What are others opinions? cc: @lamxTyler @lzmhhh123 @eurekaka

Reply:
I have a question, is the key info always not changed during the transformation rule?

Reply:
Yes, the schema is read-only. If we want to change the schema, we need to create a new Group.

Reply:
I mean we have applied some rules during optimization. Such as push down aggregation like `firstRow`, should the key info to be changed?

Reply:
Lazy build +1